### ISSUE TITLE
NAN prediction in example while training with matrix factorization

### ISSUE THREAD

**Issue by peterhurford** [JohnLangford/vowpal_wabbit#1203]
Whenever I run the following model:

```
vw -d inputs.txt --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 -c -f m.model
```

creates the following error output:

```
creating quadratic features for pairs: cp
using l2 regularization = 0.01
final_regressor = m.model
Num weight bits = 18
learning rate = 0.015
initial_t = 1
power_t = 0
decay_learning_rate = 0.97
using cache_file = inputs.txt.cache
ignoring text input in favor of cache input
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.768504 0.768504            1            1.0   1.0000   0.1234       23
0.780096 0.791688            2            2.0   1.0000   0.1102       23
0.727856 0.675615            4            4.0   1.0000   0.2065       23
9.063752 17.399648            8            8.0   1.0000   0.2300       23
15.966624 22.869497           16           16.0  13.0000   0.4051       23
69.786593 123.606562           32           32.0   4.0000   1.4780       23
39.727003 9.667413           64           64.0   1.0000   1.7759       23
29.505255 19.283507          128          128.0   6.0000   2.5476       23
35.357054 41.208854          256          256.0  11.0000   3.6256       23
27.264047 19.171039          512          512.0   3.0000   3.2394       23
24.564501 21.864955         1024         1024.0   2.0000   2.6770       23
28.429097 32.293692         2048         2048.0   2.0000   3.3020       23
5970.733917 11913.038738         4096         4096.0  15.0000   0.0000       23
NAN prediction in example 4548, forcing 0.000000
NAN prediction in example 4549, forcing 0.000000
NAN prediction in example 4550, forcing 0.000000
NAN prediction in example 4551, forcing 0.000000
NAN prediction in example 4552, forcing 0.000000
NAN prediction in example 4553, forcing 0.000000
NAN prediction in example 4554, forcing 0.000000
NAN prediction in example 4555, forcing 0.000000
NAN prediction in example 4557, forcing 0.000000
NAN prediction in example 4558, forcing 0.000000
NAN prediction in example 4561, forcing 0.000000
NAN prediction in example 4562, forcing 0.000000
NAN prediction in example 4563, forcing 0.000000
...
NAN prediction in example 32387, forcing 0.000000
NAN prediction in example 32388, forcing 0.000000
NAN prediction in example 32389, forcing 0.000000
NAN prediction in example 32390, forcing 0.000000
NAN prediction in example 32391, forcing 0.000000
NAN prediction in example 32392, forcing 0.000000
NAN prediction in example 32392, forcing 0.000000
NAN prediction in example 32393, forcing 0.000000
NAN prediction in example 32394, forcing 0.000000
NAN prediction in example 32395, forcing 0.000000
NAN prediction in example 32396, forcing 0.000000
NAN prediction in example 32397, forcing 0.000000
NAN prediction in example 32398, forcing 0.000000
NAN prediction in example 32399, forcing 0.000000
NAN prediction in example 32400, forcing 0.000000
NAN prediction in example 32401, forcing 0.000000

finished run
number of examples per pass = 5400
passes used = 6
weighted example sum = 32400.000000
weighted label sum = 134496.000000
average loss = 172.611664 h
best constant = 4.151111
total feature number = 745200
```

`inputs.txt` is here: http://pastebin.com/xsbjnR2n

I'm using a Mac OS X 10.11.6 but I can replicate this bug on an EC2 running Ubuntu.

This problem seems input-sensitive, as running with different sizes changes the results.

This fails:

```
rm m.*; head -n 4000 inputs.txt |  vw --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 --cache_file m.cache -f m.model
```

Also shuffling it still fails:

```
rm m.*; head -n 4000 inputs.txt |  gshuf | vw --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 --cache_file m.cache -f m.model
```

This works:

```
rm m.*; head -n 2000 inputs.txt |  vw --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 --cache_file m.cache -f m.model
```

This works:

```
rm m.*; tail -n 2000 inputs.txt |  vw --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 --cache_file m.cache -f m.model
```

This fails:

```
rm m.*; tail -n 4000 inputs.txt |  vw --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 --cache_file m.cache -f m.model
```

This is also a problem with the matrix factorization, as removing `--rank` fixes the issue.

This also seems to be a problem with multi-pass, as setting `passes` to 1 fixes the issue.

This is also hyperparameter sensitive, as setting `power_t` to 0.5 instead of 0.0 fixes the issue. (Though a `power_t` of 0.1 still fails.)

Lastly, this is a problem with the size of my labels, as dividing all the labels by 10 works:

```
rm m.*; cat inputs.txt | awk '{printf "%.2f %s %d %s %d\n", $1 / 10, $2, $3, $4, $5}' | vw -b 21 --learning_rate 0.015 --l2 0.01 -q cp --power_t 0.0 --decay_learning_rate 0.97 --rank 10 --passes 10 --cache_file m.cache -f m.model
```

**Comment by arielf** [comment]
I feel this is not so much a bug, since `vw` is warning loudly that it fails to converge.

The failure to converge is due to 1) the specific data (especially the label distribution) and 2) the forcing of aggressive (far from default) command-line-options.

Note that the progress output the average loss is growing very fast so pretty soon `vw` estimate of the error overflows (failure to converge).

Looking at this data:
The distribution of labels is a bit extreme.  Most of the labels (over 98% of them) have a value between 1 and 20.  Then, there are few labels with a somewhat larger value (up to ~100), and then there are a few more outliers, with huge gaps between them, one of them pretty extreme. These outlier labels are: 132, 215, 274, 374, and  2769

2769^2 is greater than 7 million, so you can see where the huge error comes from.

The problem becomes worse with the choice of aggressive, far from the default, options: most notably `--power_t 0` and extremely low `--learning_rate 0.015` . This forces `vw` to both learn very slowly and forget whatever it learned fast.

Normally when hitting the outliers, `vw` would be increasingly immune to their effect, but in this case it is forced to way over-correct in the update of each outlier example, and end up on a divergence path.

I suspect that by using only half of the data, you're simply dropping some of the outliers, and simply stopping earlier, avoiding the overflow condition.  Similarly, by dividing the labels by 10, you're delaying the overflow condition. I suspect that dividing by 10 wouldn't be enough on a larger data set as the divergence is growing exponentially with more examples, as you learned when you added multiple passes.

As you found out, any/all of the following can alleviate the problem (they are non-exclusive):

- map the outliers to lower values, closer to a max of 100 (e.g. collapse the gaps)
- alternatively, transform the labels with log(), and/or
- Use a `--loss_function quantile` (the default, `--loss_function squared` leads to larger magnitude errors) and/or
- be less aggressive on `--power_t`
- be less aggressive on lowering `--learning_rate`

Short summary: make sure your `mean loss` column generally converges over time. i.e. you're actually learning.  Whenever you see exponential growth in error and warnings from `vw` stop, and fix your data and/or command line to avoid this.

**Comment by peterhurford** [comment]
Thanks @arielf for the detailed answer. I really appreciate it and it helps me learn more about how VW works. At this point, I would consider the matter resolved and would be comfortable with the issue being closed.

If you have more time, can you elaborate a bit on how the estimate of the error overflows and how that indicates a failure to converge? I assume there's some sort of maximum value the error can take? And that failure to converge means that the error is increasing over time rather than decreasing?

Also, if I could suggest a change, "NAN prediction in example" is a bit confusing to encounter during model training. Usually prediction implies the error is happening during prediction (which I first thought and debugged the wrong part) and NAN implies that a NAN input was given (I've encountered this error before by having a misformatted VW string), but neither were the case this time. Some message that indicates it's an overflow or a failure to converge would be more helpful, though this would likely still cause me to debug based on the bit rate (which I also did try).

Ultimately, though, I agree that it's a very weird edge case that likely no one else will encounter, so I'm fine just leaving it here.

**Comment by peterhurford** [comment]
Also @arielf just in case you're curious how I strayed so far from the defaults, the parameters I used were taken naïvely from the Matrix Factorization example in the Wiki: https://github.com/JohnLangford/vowpal_wabbit/wiki/Matrix-factorization-example

Though I completely agree that the DV we chose is very different than the MovieLens data set and are probably not appropriate for matrix factorization.

**Comment by arielf** [comment]
Trying to answer the question(s) from the very basic.  Apologies if this is too basic:

- "Learning" means there's progress towards a stable model. This is equivalent to the error (aka loss) decreasing over time.
- A smaller error (first, leftmost column in `vw` progress output) over time is the same as saying the model is converging.
- The convergence process relies on floating-point (FP) math: in each example, the actual (aka current) label (5th column in `vw` progress output) is compared with the model prediction (6th column). The difference (actual minus predicted value) between the two, is monotonic with the error of the model.  By default, (`--loss_function squared`),  half of the square of this difference is used as the error estimate.  This error estimate is used in the "update" step which adjusts the weights of the variables in the mode before proceeding to the next example.  See: [loss functions in the wiki](https://github.com/JohnLangford/vowpal_wabbit/wiki/Loss-functions)
- Computing the error and updating the weights both involve FP operations.  When an FP operation yields a NaN (Not a Number), `vw` can't update the weights, i.e. can't learn.
- Several FP operations can lead to a NaN.  You can find [a list of them in Wikipedia](https://en.wikipedia.org/wiki/NaN)

I suspect there's some overflow because I see the error growing exponentially but it can be also an underflow (e.g. when dividing by a too small number yielding "infinity")

Disclaimer: I'm not an expert on FP implementations.

HTH
